frontend/mux.go

```diff
@@ -0,0 +1,515 @@
+	return &noSuchHandlerError{Target: target, Available: available}
+}
+
+func (err *noSuchHandlerError) Error() string {
+	return fmt.Sprintf("no such handler for target %q: available targets: %s", err.Target, strings.Join(err.Available, ", "))
+}
+
+func (m *BuildMux) lookupTarget(ctx context.Context, target string) (matchedPattern string, _ *handler, _ error) {
+	// `target` is from `docker build --target=<target>`
+	// cases for `t` are as follows:
+	//    1. may have an exact match in the handlers (ideal)
+	//    2. may have a prefix match in the handlers, e.g. hander for `foo`, `target == "foo/bar"` (assume nested route)
+	// 	  3. No matching handler and `target == ""` and there is a default handler set (assume default handler)
+	//    4. No match in the handlers (error)
+	h, ok := m.handlers[target]
+	if ok {
+		return target, &h, nil
+	}
+
+	if target == "" && m.defaultH != nil {
+		bklog.G(ctx).Info("Using default target")
+		return target, m.defaultH, nil
+	}
+
+	for k, h := range m.handlers {
+		if strings.HasPrefix(target, k+"/") {
+			bklog.G(ctx).WithField("prefix", k).WithField("matching request", target).Info("Using prefix match for target")
+			return k, &h, nil
+		}
+	}
+
+	return "", nil, handlerNotFound(target, maps.Keys(m.handlers))
+}
+
+// Handle is a [gwclient.BuildFunc] that routes requests to registered handlers
+func (m *BuildMux) Handle(ctx context.Context, client gwclient.Client) (_ *gwclient.Result, retErr error) {
+	// Cache the opts in case this is the raw client
+	// This prevents a grpc request for multiple calls to BuildOpts
+	opts := client.BuildOpts().Opts
+	origOpts := dalec.DuplicateMap(opts)
+
+	t := opts[keyTarget]
+
+	defer func() {
+		if retErr != nil {
+			if _, ok := origOpts[keyTargetOpt]; !ok {
+				retErr = errors.Wrapf(retErr, "error handling requested build target %q", t)
+
+				// If we have a spec name, load it to make the error message more helpful
+				spec, _ := m.loadSpec(ctx, client)
+				if spec != nil && spec.Name != "" {
+					retErr = errors.Wrapf(retErr, "spec: %s", spec.Name)
+				}
+			}
+		}
+	}()
+
+	ctx = bklog.WithLogger(ctx, bklog.G(ctx).WithField("handlers", maps.Keys(m.handlers)).WithField("target", opts[keyTarget]).WithField("requestid", opts[requestIDKey]).WithField("targetKey", GetTargetKey(client)))
+
+	bklog.G(ctx).Info("Handling request")
+
+	res, handled, err := m.handleSubrequest(ctx, client, opts)
+	if err != nil {
+		return nil, err
+	}
+	if handled {
+		return res, nil
+	}
+
+	matched, h, err := m.lookupTarget(ctx, t)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = bklog.WithLogger(ctx, bklog.G(ctx).WithField("matched", matched))
+
+	// each call to `Handle` handles the next part of the target
+	// When we call the handler, we want to remove the part of the target that is being handled so the next handler can handle the next part
+	client = maybeSetDalecTargetKey(trimTargetOpt(client, matched), matched)
```

Comment:
nit: un-nest the call to `trimTargetOpt`. It's an important mutation of state and it's too easy to miss when it's nested.

Reply:
Fixed.